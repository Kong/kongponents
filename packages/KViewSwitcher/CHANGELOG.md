# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 6.1.10 (2021-11-18)

**Note:** Version bump only for package @kongponents/kviewswitcher





## 6.1.9 (2021-11-18)

**Note:** Version bump only for package @kongponents/kviewswitcher





## 6.1.8 (2021-11-18)


### Bug Fixes

* **css:** box-shadow does not render on ios ([#420](https://github.com/Kong/kongponents/issues/420)) ([c0ba7cb](https://github.com/Kong/kongponents/commit/c0ba7cb90bf3c158a1baa6b879d33230359db830))





## 6.1.7 (2021-11-17)

**Note:** Version bump only for package @kongponents/kviewswitcher





## 6.1.6 (2021-11-17)


### Bug Fixes

* **ktextarea:** limit exceeded event ([#418](https://github.com/Kong/kongponents/issues/418)) ([f7cf421](https://github.com/Kong/kongponents/commit/f7cf421a4ded7f5b99eaa5a18d02a27d450020b7))





## 6.1.5 (2021-11-17)

**Note:** Version bump only for package @kongponents/kviewswitcher





## 6.1.4 (2021-11-17)


### Bug Fixes

* general fixes ([#415](https://github.com/Kong/kongponents/issues/415)) ([5b4bb48](https://github.com/Kong/kongponents/commit/5b4bb484d74e75779d39cf901b2cdaa23818e8e7))





## 6.1.3 (2021-11-17)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [6.1.2](https://github.com/Kong/kongponents/compare/@kongponents/kviewswitcher@5.0.8...@kongponents/kviewswitcher@6.1.2) (2021-11-17)

Kongponents redesign - no changes to this component


## [5.0.8](https://github.com/Kong/kongponents/compare/v5.0.7...v5.0.8) (2021-11-15)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [5.0.7](https://github.com/Kong/kongponents/compare/v5.0.6...v5.0.7) (2021-10-26)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [5.0.6](https://github.com/Kong/kongponents/compare/v5.0.5...v5.0.6) (2021-10-26)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [5.0.5](https://github.com/Kong/kongponents/compare/v5.0.4...v5.0.5) (2021-10-13)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [5.0.4](https://github.com/Kong/kongponents/compare/v5.0.3...v5.0.4) (2021-10-06)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [5.0.3](https://github.com/Kong/kongponents/compare/v5.0.2...v5.0.3) (2021-09-22)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [5.0.2](https://github.com/Kong/kongponents/compare/v5.0.1...v5.0.2) (2021-09-15)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [5.0.1](https://github.com/Kong/kongponents/compare/v5.0.0...v5.0.1) (2021-09-13)

**Note:** Version bump only for package @kongponents/kviewswitcher





# [5.0.0](https://github.com/Kong/kongponents/compare/v4.0.2...v5.0.0) (2021-09-13)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [4.0.2](https://github.com/Kong/kongponents/compare/v4.0.1...v4.0.2) (2021-09-09)

**Note:** Version bump only for package @kongponents/kviewswitcher





## [4.0.1](https://github.com/Kong/kongponents/compare/v4.0.0...v4.0.1) (2021-09-07)

**Note:** Version bump only for package @kongponents/kviewswitcher





# 4.0.0 (2021-09-03)

**Note:** Version bump only for package @kongponents/kviewswitcher





## 0.1.1 (2020-11-17)

**Note:** Version bump only for package @kongponents/kviewswitcher
