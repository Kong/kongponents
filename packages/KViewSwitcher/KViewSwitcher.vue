<template>
  <KButton
    :class="[view, { 'is-animating': isAnimating }]"
    class="view-switch-button"
    @click="toggleView"
  >
    <div class="icon">
      <div class="dots">
        <i
          v-for="i in 4"
          :key="i" />
      </div>
      <div class="lines">
        <i
          v-for="i in 4"
          :key="i" />
      </div>
    </div>
    <div
      v-if="hasLabel"
      class="text">
      <span>Grid</span>
      <span>List</span>
    </div>
  </KButton>
</template>

<script>
import KButton from '@kongponents/kbutton/KButton.vue'

export default {
  name: 'ViewSwitcher',
  components: { KButton },
  props: {
    view: {
      type: String,
      required: true,
      validator: (val) => ['grid', 'list'].includes(val)
    },
    hasLabel: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      isAnimating: false
    }
  },
  methods: {
    toggleView () {
      this.isAnimating = true // can stay on after initial click
      setTimeout(() => {
        // this.isAnimating = false// can stay on after initial click
      }, 500)
      this.$emit('view-changed', this.view === 'list' ? 'grid' : 'list')
    }
  }
}
</script>

<style lang="scss">
// Originally forked and modified from https://codepen.io/aaroniker/pen/dyoKeMP
.view-switch-button {
  --KButtonSecondaryBackground: var(--white);
  --KButtonSecondaryActiveBackground: var(--white);
  --KButtonPaddingY: var(--spacing-xxs);
  --KButtonPaddingX: var(--spacing-xxs);

  --color: var(--black-70);
  --icon-color: var(--black-25);
  // --span-name: text;
  // --dots-name: back;
  // --lines-name: scale-down;
  padding: 6px 12px 6px 8px;
  margin: 0;
  display: flex;
  outline: none;
  position: relative;
  border: none;
  border-radius: 9px;
  background: var(--b, var(--background));
  cursor: pointer;
  -webkit-appearence: none;
  -webkit-tap-highlight-color: transparent;
  .icon {
    width: 24px;
    height: 24px;
    position: relative;
    i {
      position: absolute;
      left: var(--left, 4px);
      top: var(--top, 4px);
      display: block;
      border-radius: 2px;
      width: var(--width, 7px);
      height: var(--height, 7px);
      background: var(--icon-color);
      animation: var(--name, var(--dots-name, none))
        var(--duration, var(--dots-duration, 5s))
        var(--easing, var(--dots-easing, linear)) forwards
        var(--delay, var(--dots-delay, 0s));
    }
    .dots {
      i {
        &:nth-child(1) {
          --x-middle: -8px;
          --y-middle: 10px;
          --x-end: -2px;
          --y-end: 12px;
          --x-back: 10px;
          --y-back: 7px;
          --x-back-end: 9px;
          --y-back-end: 0;
        }
        &:nth-child(2) {
          --left: 13px;
          --x-middle: -12px;
          --y-middle: 5px;
          --x-end: -11px;
          --y-end: 7px;
          --x-back: -3px;
          --y-back: 1px;
          --x-back-end: -9px;
          --y-back-end: 0;
        }
        &:nth-child(3) {
          --top: 13px;
          --x-middle: 4px;
          --y-middle: -5px;
          --x-end: -2px;
          --y-end: -7px;
          --x-back: -5px;
          --y-back: 0px;
          --x-back-end: 9px;
          --y-back-end: 0;
        }
        &:nth-child(4) {
          --left: 13px;
          --top: 13px;
          --x-middle: 0;
          --y-middle: -10px;
          --x-end: -11px;
          --y-end: -12px;
          --x-back: -14px;
          --y-back: -8px;
          --x-back-end: -9px;
          --y-back-end: 0;
        }
      }
    }
    .lines {
      --name: var(--lines-name, none);
      --duration: var(--lines-duration, 0.15s);
      --easing: var(--lines-easing, linear);
      --delay: var(--lines-delay, 0s);
      i {
        --left: 9px;
        --top: 3px;
        --height: 2px;
        --width: 11px;
        transform-origin: 0 50%;
        transform: translateY(20%) translateZ(0) scaleX(0);
        &:nth-child(2) {
          --top: 8px;
        }
        &:nth-child(3) {
          --top: 13px;
        }
        &:nth-child(4) {
          --top: 18px;
        }
        &:nth-child(3),
        &:nth-child(4) {
          transform-origin: 100% 50%;
        }
      }
    }
  }
  .text {
    margin-left: 4px;
    position: relative;
    line-height: 24px;
    font-weight: 600;
    font-size: 14px;
    min-width: 28px;
    color: var(--color);
    span {
      --y-default: 0;
      --o-default: 1;
      --y-active: -12px;
      --o-active: 0;
      display: block;
      opacity: var(--o-default);
      transform: translateY(var(--y-default)) translateZ(0);
      animation: var(--span-name, none) 0.4s ease forwards;
      &:last-child {
        --y-default: 12px;
        --o-default: 0;
        --y-active: 0;
        --o-active: 1;
        position: absolute;
        left: 0;
        top: 0;
      }
    }
  }
  &.list {
    // animation: none;
    --span-name: text-active;
    --dots-name: none;
    --lines-name: scale;
    .dots i {
      // animation: none;
      // transform: translate(var(--x-end, 0), var(--y-end, 0)) scale(0.4);
    }
    .lines {
     i {
      //  animation: none;
        // transform: translateY(20%) translateZ(0) scaleX(1);
      }
    }
  }
  &.is-animating {
    --span-name: text;
    --dots-name: back;
    --lines-name: scale-down;
    .lines {
      i {
        transform-origin: 0 50%;
        &:nth-child(3),
        &:nth-child(4) {
          transform-origin: 100% 50%;
        }
      }
    }
    &.list {
      --span-name: text-active;
      --dots-name: move;
      --lines-name: scale;
      --lines-duration: .15s;
      --lines-delay: .3s;
      .lines {
        i {
          transform-origin: 100% 50%;
          &:nth-child(3),
          &:nth-child(4) {
            transform-origin: 0 50%;
          }
        }
      }
    }
  }
}

@keyframes text {
  0% {
    opacity: var(--o-active);
    transform: translateY(var(--y-active)) translateZ(0);
  }
  100% {
    opacity: var(--o-default);
    transform: translateY(var(--y-default)) translateZ(0);
  }
}

@keyframes text-active {
  0% {
    opacity: var(--o-default);
    transform: translateY(var(--y-default)) translateZ(0);
  }
  100% {
    opacity: var(--o-active);
    transform: translateY(var(--y-active)) translateZ(0);
  }
}

@keyframes move {
  50% {
    transform: translate(var(--x-middle, 0), var(--y-middle, 0)) scale(0.4);
  }
  100% {
    transform: translate(var(--x-end, 0), var(--y-end, 0)) scale(0.4);
  }
}

@keyframes back {
  0%,
  15% {
    transform: translate(var(--x-end, 0), var(--y-end, 0)) scale(0.4);
  }
  50% {
    transform: translate(var(--x-back, 0), var(--y-back, 0)) scale(0.5);
  }
  100% {
    transform: translate(var(--x-back-end, 0), var(--y-back-end, 0)) scale(1);
  }
}

@keyframes scale {
  100% {
    transform: translateY(20%) translateZ(0) scaleX(1);
  }
}

@keyframes scale-down {
  0% {
    transform: translateY(20%) translateZ(0) scaleX(1);
  }
  100% {
    transform: translateY(20%) translateZ(0) scaleX(0);
  }
}
</style>
