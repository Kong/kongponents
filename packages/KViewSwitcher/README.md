# @kongponents/kviewswitcher

[![](https://img.shields.io/npm/v/@kongponents/kviewswitcher.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kviewswitcher)

```vue
<KViewSwitcher :description="'hello world'">
  Hello from a slot
</KViewSwitcher>
```
