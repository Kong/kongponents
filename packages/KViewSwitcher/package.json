{
  "name": "@kongponents/kviewswitcher",
  "version": "6.1.17",
  "description": "KViewSwitcher description here.",
  "main": "dist/KViewSwitcher.umd.min.js",
  "componentName": "KViewSwitcher",
  "source": "KViewSwitcher.vue",
  "files": [
    "dist",
    "KViewSwitcher.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  }
}
