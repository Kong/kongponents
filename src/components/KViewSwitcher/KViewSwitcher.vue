
<template>
  <KButton
    appearance="outline"
    class="k-view-switcher view-switch-button non-visual-button"
    :class="[view, { paused: isPaused }]"
    :is-rounded="false"
    size="small"
    :title="`Toggle to ${view === 'table' ? 'grid' : 'table'} view`"
    @click="toggleView"
  >
    <div class="icon">
      <div class="dots">
        <i
          v-for="i in 4"
          :key="i"
        />
      </div>
      <div class="lines">
        <i
          v-for="i in 4"
          :key="i"
        />
      </div>
    </div>
  </KButton>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue'
import KButton from '@/components/KButton/KButton.vue'

export default defineComponent({
  name: 'KViewSwitcher',
  components: { KButton },
  props: {
    view: {
      type: String,
      default: 'table',
      required: true,
      validator: (val: string): boolean => ['table', 'grid'].includes(val),
    },
  },
  emits: ['view-changed'],
  setup(props, { emit }) {
    const isPaused = ref(true)

    const toggleView = (): void => {
      // Removed paused animations
      isPaused.value = false

      // Emit new view
      emit('view-changed', props.view === 'table' ? 'grid' : 'table')
    }

    return {
      isPaused,
      toggleView,
    }
  },
})
</script>

<style lang="scss" scoped>
@import '@/styles/variables';
@import '@/styles/functions';

// Originally forked and modified from https://codepen.io/aaroniker/pen/dyoKeMP
.view-switch-button {
  --KButtonPaddingY: 6px;
  --KButtonPaddingX: 6px;
  --KButtonSecondaryHover: var(--white);
  --KButtonSecondaryHoverBorder: var(--blue-300);
  --KButtonSecondaryFocus: none;
  transform: scale(var(--scale, 1)) translateZ(0);

  &.paused .icon i {
    animation-duration: 0s;
  }

  .icon {
    position: relative;
    width: 24px;
    height: 24px;

    i {
      position: absolute;
      top: var(--top, 4px);
      left: var(--left, 4px);
      display: block;
      width: var(--width, 7px);
      height: var(--height, 7px);
      background-color: var(--grey-500);
      border-radius: 2px;
      transition: background-color 200ms ease;
      animation: var(--name, var(--dots-name, none)) var(--duration, var(--dots-duration, .5s)) var(--easing, var(--dots-easing, linear)) forwards var(--delay, var(--dots-delay, 0s));
    }

    .dots i {
      &:nth-child(1) {
        --x-middle: -8px;
        --y-middle: 10px;
        --x-end: -2px;
        --y-end: 12px;
        --x-back: 10px;
        --y-back: 7px;
        --x-back-end: 9px;
        --y-back-end: 0;
      }

      &:nth-child(2) {
        --left: 13px;
        --x-middle: -12px;
        --y-middle: 5px;
        --x-end: -11px;
        --y-end: 7px;
        --x-back: -3px;
        --y-back: 1px;
        --x-back-end: -9px;
        --y-back-end: 0;
      }

      &:nth-child(3) {
        --top: 13px;
        --x-middle: 4px;
        --y-middle: -5px;
        --x-end: -2px;
        --y-end: -7px;
        --x-back: -5px;
        --y-back: 0px;
        --x-back-end: 9px;
        --y-back-end: 0;
      }

      &:nth-child(4) {
        --left: 13px;
        --top: 13px;
        --x-middle: 0;
        --y-middle: -10px;
        --x-end: -11px;
        --y-end: -12px;
        --x-back: -14px;
        --y-back: -8px;
        --x-back-end: -9px;
        --y-back-end: 0;
      }
    }

    .lines {
      --name: var(--lines-name, none);
      --duration: var(--lines-duration, 0.15s);
      --easing: var(--lines-easing, linear);
      --delay: var(--lines-delay, 0s);

      i {
        --left: 9px;
        --top: 3px;
        --height: 2px;
        --width: 11px;
        transform: translateY(20%) translateZ(0) scaleX(0);
        transform-origin: 0 50%;
        &:nth-child(2) { --top: 8px; }
        &:nth-child(3) { --top: 13px; }
        &:nth-child(4) { --top: 18px; }
        &:nth-child(3),
        &:nth-child(4) { transform-origin: 100% 50%; }
      }
    }
  }

  &.table {
    --dots-name: back;
    --lines-name: scale-down;

    .lines i {
      transform-origin: 0 50%;
      &:nth-child(3),
      &:nth-child(4) { transform-origin: 100% 50%; }
    }
  }

  &.grid {
    --dots-name: move;
    --lines-name: scale;
    --lines-duration: 0.15s;
    --lines-delay: 0.3s;

    .lines i {
      transform-origin: 100% 50%;
      &:nth-child(3),
      &:nth-child(4) { transform-origin: 0 50%; }
    }
  }
}
</style>

<style lang="scss">
// @keyframes animations need to be un-scoped

.k-view-switcher {
  @keyframes move {
    50% { transform: translate(var(--x-middle, 0), var(--y-middle, 0)) scale(.4); }
    100% { transform: translate(var(--x-end, 0), var(--y-end, 0)) scale(.4); }
  }

  @keyframes back {
    0%,
    15% { transform: translate(var(--x-end, 0), var(--y-end, 0)) scale(.4); }
    50% { transform: translate(var(--x-back, 0), var(--y-back, 0)) scale(.5); }
    100% { transform: translate(var(--x-back-end, 0), var(--y-back-end, 0)) scale(1); }
  }

  @keyframes scale {
    100% { transform: translateY(20%) translateZ(0) scaleX(1); }
  }

  @keyframes scale-down {
    0% { transform: translateY(20%) translateZ(0) scaleX(1); }
    100% { transform: translateY(20%) translateZ(0) scaleX(0); }
  }
}
</style>
