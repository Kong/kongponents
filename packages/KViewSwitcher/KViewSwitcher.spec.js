import { mount } from '@vue/test-utils'
import KViewSwitcher from '@/KViewSwitcher/KViewSwitcher'

describe('KViewSwitcher', () => {
  it('matches snapshot', () => {
    const wrapper = mount(KViewSwitcher)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
