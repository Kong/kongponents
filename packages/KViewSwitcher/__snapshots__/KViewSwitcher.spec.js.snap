// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KViewSwitcher matches snapshot 1`] = `
<button type="button" class="k-button view-switch-button small outline table paused" title="Toggle to grid view">
  <!---->
  <div class="icon">
    <div class="dots"><i></i><i></i><i></i><i></i></div>
    <div class="lines"><i></i><i></i><i></i><i></i></div>
  </div>
  <!---->
</button>
`;
